File: frontend/src/utils/guid.ts
/**
 * GUID utilities for entity identification.
 *
 * Provides validation and parsing functions for GUIDs
 * in the format {prefix}_{base32_uuid}.
 *
 * GUID Format:
 *   - prefix: 3-character entity type identifier
 *   - separator: underscore (_)
 *   - uuid: 26-character Crockford Base32 encoded UUIDv7
 *
 * Entity Prefixes:
 *   Database entities (persisted):
 *   - col: Collection
 *   - con: Connector
 *   - pip: Pipeline
 *   - res: AnalysisResult
 *   - evt: Event (calendar event)
 *   - ser: EventSeries (multi-day event grouping)
 *   - loc: Location (known locations)
 *   - org: Organizer (event organizers)
 *   - prf: Performer (event performers)
 *   - cat: Category (event categories)
 *   In-memory entities (transient):
 *   - job: Job
 *   - imp: Import
 */

/**
 * Entity type prefixes
 */
export type EntityPrefix =
  | 'col'
  | 'con'
  | 'pip'
  | 'res'
  | 'evt'
  | 'ser'
  | 'loc'
  | 'org'
  | 'prf'
  | 'cat'
  | 'job'
  | 'imp'

/**
 * Entity type names mapped to prefixes
 */
export const ENTITY_PREFIXES: Record<EntityPrefix, string> = {
  col: 'Collection',
  con: 'Connector',
  pip: 'Pipeline',
  res: 'AnalysisResult',
  evt: 'Event',
  ser: 'EventSeries',
  loc: 'Location',
  org: 'Organizer',
  prf: 'Performer',
  cat: 'Category',
  job: 'Job',
  imp: 'Import',
}

/**
 * Crockford Base32 alphabet (excludes I, L, O, U to avoid confusion)
 */
const CROCKFORD_ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'

/**
 * Regex pattern for validating GUIDs
 * Format: {3-char prefix}_{26-char Crockford Base32}
 */
const GUID_PATTERN =
  /^(col|con|pip|res|evt|ser|loc|org|prf|cat|job|imp)_[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/i

/**
 * Validates if a string is a valid GUID.
 *
 * @param id - The string to validate
 * @param expectedPrefix - Optional prefix to validate against
 * @returns True if valid GUID format
 *
 * @example
 * isValidGuid('col_01hgw2bbg0000000000000000') // true
 * isValidGuid('col_01hgw2bbg0000000000000000', 'col') // true
 * isValidGuid('col_01hgw2bbg0000000000000000', 'con') // false
 */
export function isValidGuid(id: string, expectedPrefix?: EntityPrefix): boolean {
  if (!id || typeof id !== 'string') {
    return false
  }

  if (!GUID_PATTERN.test(id)) {
    return false
  }

  if (expectedPrefix) {
    const prefix = id.slice(0, 3).toLowerCase() as EntityPrefix
    return prefix === expectedPrefix
  }

  return true
}

/**
 * Extracts the entity type from a GUID.
 *
 * @param guid - The GUID string
 * @returns The entity type name, or null if invalid
 *
 * @example
 * getEntityType('col_01hgw2bbg0000000000000000') // 'Collection'
 * getEntityType('con_01hgw2bbg0000000000000001') // 'Connector'
 * getEntityType('invalid') // null
 */
export function getEntityType(guid: string): string | null {
  if (!guid || guid.length < 3) {
    return null
  }

  const prefix = guid.slice(0, 3).toLowerCase() as EntityPrefix
  return ENTITY_PREFIXES[prefix] || null
}

/**
 * Extracts the prefix from a GUID.
 *
 * @param guid - The GUID string
 * @returns The prefix, or null if invalid
 *
 * @example
 * getPrefix('col_01hgw2bbg0000000000000000') // 'col'
 * getPrefix('invalid') // null
 */
export function getPrefix(guid: string): EntityPrefix | null {
  if (!isValidGuid(guid)) {
    return null
  }

  return guid.slice(0, 3).toLowerCase() as EntityPrefix
}

/**
 * Checks if a string is a numeric ID (for backward compatibility).
 *
 * @param id - The string to check
 * @returns True if the string is numeric
 *
 * @example
 * isNumericId('123') // true
 * isNumericId('col_xxx') // false
 */
export function isNumericId(id: string): boolean {
  if (!id || typeof id !== 'string') {
    return false
  }

  return /^\d+$/.test(id)
}

/**
 * Checks if a string is a GUID.
 *
 * @param id - The string to check
 * @returns True if the string matches GUID format
 *
 * @example
 * isGuid('col_01hgw2bbg0000000000000000') // true
 * isGuid('123') // false
 */
export function isGuid(id: string): boolean {
  return isValidGuid(id)
}

/**
 * Determines the identifier type (numeric or guid).
 *
 * @param id - The identifier string
 * @returns 'numeric' | 'guid' | 'invalid'
 *
 * @example
 * getIdentifierType('123') // 'numeric'
 * getIdentifierType('col_xxx') // 'guid'
 * getIdentifierType('invalid') // 'invalid'
 */
export function getIdentifierType(id: string): 'numeric' | 'guid' | 'invalid' {
  if (isNumericId(id)) {
    return 'numeric'
  }

  if (isGuid(id)) {
    return 'guid'
  }

  return 'invalid'
}

/**
 * Formats a GUID for display (truncated with ellipsis).
 *
 * @param guid - The GUID string
 * @param showPrefix - Whether to include the prefix (default: true)
 * @returns Formatted string for display
 *
 * @example
 * formatGuid('col_01hgw2bbg0000000000000000') // 'col_01hgw2bb...'
 * formatGuid('col_01hgw2bbg0000000000000000', false) // '01hgw2bb...'
 */
export function formatGuid(guid: string, showPrefix = true): string {
  if (!isValidGuid(guid)) {
    return guid
  }

  if (showPrefix) {
    // Show prefix + first 8 chars of base32 + ellipsis
    return guid.slice(0, 12) + '...'
  }

  // Show first 8 chars of base32 + ellipsis
  return guid.slice(4, 12) + '...'
}

/**
 * Copies a GUID to the clipboard.
 *
 * @param guid - The GUID to copy
 * @returns Promise that resolves when copied
 */
export async function copyGuid(guid: string): Promise<void> {
  if (typeof navigator?.clipboard?.writeText === 'function') {
    await navigator.clipboard.writeText(guid)
  } else {
    // Fallback for older browsers
    const textArea = document.createElement('textarea')
    textArea.value = guid
    textArea.style.position = 'fixed'
    textArea.style.opacity = '0'
    document.body.appendChild(textArea)
    textArea.select()
    document.execCommand('copy')
    document.body.removeChild(textArea)
  }
}

/**
 * Validates a GUID and returns it if valid, throws if invalid.
 * Use this before using a GUID in API URLs to prevent URL injection.
 *
 * @param guid - The GUID to validate
 * @param expectedPrefix - Optional prefix to validate against
 * @returns The validated GUID
 * @throws Error if GUID is invalid
 *
 * @example
 * const validGuid = validateGuid(userInput, 'res')
 * await api.get(`/results/${validGuid}`)
 */
export function validateGuid(guid: string, expectedPrefix?: EntityPrefix): string {
  if (!isValidGuid(guid, expectedPrefix)) {
    const prefixMsg = expectedPrefix ? ` with prefix '${expectedPrefix}'` : ''
    throw new Error(`Invalid GUID format${prefixMsg}: ${guid}`)
  }
  return guid
}

File: backend/src/services/guid.py
"""
GUID service for entity identification.

Provides utilities for generating, encoding, decoding, and validating
Global Unique Identifiers used in URLs and API responses.

GUID Format: {prefix}_{base32_uuid}
- prefix: 3-character entity type identifier (col, con, pip, res)
- base32_uuid: 26-character Crockford Base32 encoded UUIDv7
"""

import re
import uuid

import base32_crockford
from uuid_extensions import uuid7

# Prefix mappings for entity types
# Database entities (persisted):
#   col - Collection
#   con - Connector
#   pip - Pipeline
#   res - AnalysisResult
#   evt - Event (calendar event)
#   ser - EventSeries (multi-day event grouping)
#   loc - Location (known locations)
#   org - Organizer (event organizers)
#   prf - Performer (event performers)
#   cat - Category (event categories)
# In-memory entities (transient):
#   job - Tool execution job
#   imp - Config import session
ENTITY_PREFIXES = {
    "col": "Collection",
    "con": "Connector",
    "pip": "Pipeline",
    "res": "AnalysisResult",
    "evt": "Event",
    "ser": "EventSeries",
    "loc": "Location",
    "org": "Organizer",
    "prf": "Performer",
    "cat": "Category",
    "job": "Job",
    "imp": "ImportSession",
}

# Crockford Base32 alphabet (excludes I, L, O, U to avoid confusion)
CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

# Pattern for validating GUIDs
# Format: {3-char prefix}_{26-char Crockford Base32}
GUID_PATTERN = re.compile(
    r"^(col|con|pip|res|evt|ser|loc|org|prf|cat|job|imp)_[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$",
    re.IGNORECASE
)


class GuidService:
    """
    Service for GUID operations.

    Provides static methods for:
    - Generating new UUIDv7 values
    - Encoding UUIDs to GUID strings
    - Decoding GUID strings to UUIDs
    - Validating GUID format
    - Parsing identifiers (distinguishing GUID vs numeric IDs)
    """

    @staticmethod
    def generate_uuid() -> uuid.UUID:
        """
        Generate a new UUIDv7 value.

        UUIDv7 is time-ordered for better database indexing performance.

        Returns:
            New UUID object
        """
        return uuid7()

    @staticmethod
    def encode_uuid(uuid_value: uuid.UUID, prefix: str) -> str:
        """
        Encode a UUID to a GUID string.

        Args:
            uuid_value: UUID to encode
            prefix: Entity type prefix (col, con, pip, res)

        Returns:
            GUID string (e.g., "col_01HGW2BBG...")

        Raises:
            ValueError: If prefix is invalid
        """
        if prefix not in ENTITY_PREFIXES:
            raise ValueError(
                f"Invalid prefix '{prefix}'. "
                f"Valid prefixes: {', '.join(ENTITY_PREFIXES.keys())}"
            )

        # Convert UUID to integer and encode as Crockford Base32
        if isinstance(uuid_value, bytes):
            uuid_int = int.from_bytes(uuid_value, "big")
        else:
            uuid_int = int.from_bytes(uuid_value.bytes, "big")

        encoded = base32_crockford.encode(uuid_int)
        # Pad to 26 characters
        encoded = encoded.zfill(26)
        return f"{prefix}_{encoded.lower()}"

    @staticmethod
    def generate_guid(prefix: str) -> str:
        """
        Generate a new GUID with the specified prefix.

        Convenience method that combines generate_uuid() and encode_uuid().

        Args:
            prefix: Entity type prefix (col, con, pip, res, job, imp)

        Returns:
            New GUID string (e.g., "job_01hgw2bbg...")

        Raises:
            ValueError: If prefix is invalid

        Example:
            >>> job_id = GuidService.generate_guid("job")
            >>> session_id = GuidService.generate_guid("imp")
        """
        return GuidService.encode_uuid(GuidService.generate_uuid(), prefix)

    @staticmethod
    def decode_guid(guid: str) -> tuple[str, uuid.UUID]:
        """
        Decode a GUID string to its components.

        Args:
            guid: GUID string (e.g., "col_01HGW2BBG...")

        Returns:
            Tuple of (prefix, UUID)

        Raises:
            ValueError: If the GUID format is invalid
        """
        if not guid:
            raise ValueError("GUID cannot be empty")

        if not GUID_PATTERN.match(guid):
            raise ValueError(
                f"Invalid GUID format: {guid}. "
                f"Expected format: {{prefix}}_{{26-char base32}}"
            )

        prefix = guid[:3].lower()
        encoded_part = guid[4:]  # Skip "xxx_"

        try:
            uuid_int = base32_crockford.decode(encoded_part.upper())
            uuid_bytes = uuid_int.to_bytes(16, "big")
            return prefix, uuid.UUID(bytes=uuid_bytes)
        except (ValueError, OverflowError) as e:
            raise ValueError(f"Invalid GUID encoding: {e}")

    @staticmethod
    def validate_guid(guid: str, expected_prefix: str = None) -> bool:
        """
        Validate a GUID format.

        Args:
            guid: GUID string to validate
            expected_prefix: Optional expected prefix for type checking

        Returns:
            True if valid, False otherwise
        """
        if not guid:
            return False

        if not GUID_PATTERN.match(guid):
            return False

        if expected_prefix:
            prefix = guid[:3].lower()
            return prefix == expected_prefix.lower()

        return True

    @staticmethod
    def get_entity_type(guid: str) -> str | None:
        """
        Get the entity type name from a GUID.

        Args:
            guid: GUID string

        Returns:
            Entity type name (Collection, Connector, etc.) or None if invalid
        """
        if not guid or len(guid) < 3:
            return None

        prefix = guid[:3].lower()
        return ENTITY_PREFIXES.get(prefix)

    @staticmethod
    def is_numeric_id(identifier: str) -> bool:
        """
        Check if an identifier is a numeric ID.

        Args:
            identifier: ID string to check

        Returns:
            True if the identifier is numeric (integer)
        """
        if not identifier:
            return False
        return identifier.isdigit()

    @staticmethod
    def is_guid(identifier: str) -> bool:
        """
        Check if an identifier is a GUID.

        Args:
            identifier: ID string to check

        Returns:
            True if the identifier matches GUID format
        """
        return GuidService.validate_guid(identifier)

    @staticmethod
    def parse_identifier(identifier: str, expected_prefix: str = None) -> uuid.UUID:
        """
        Parse a GUID identifier and return the UUID.

        Only accepts GUID format identifiers (numeric IDs are no longer supported).

        Args:
            identifier: GUID string (e.g., "col_01HGW2BBG...")
            expected_prefix: Expected prefix for validation (col, con, pip, res)

        Returns:
            UUID object extracted from the GUID

        Raises:
            ValueError: If the identifier format is invalid or prefix doesn't match
        """
        if not identifier:
            raise ValueError("Identifier cannot be empty")

        # Check if GUID
        if GUID_PATTERN.match(identifier):
            if expected_prefix:
                prefix = identifier[:3].lower()
                if prefix != expected_prefix.lower():
                    raise ValueError(
                        f"GUID prefix mismatch. "
                        f"Expected '{expected_prefix}', got '{prefix}'"
                    )

            _prefix, uuid_value = GuidService.decode_guid(identifier)
            return uuid_value

        # Provide helpful error for numeric IDs
        if identifier.isdigit():
            raise ValueError(
                f"Numeric IDs are no longer supported. "
                f"Please use GUID format ({{prefix}}_{{base32}})"
            )

        raise ValueError(
            f"Invalid identifier format: {identifier}. "
            f"Expected GUID format ({{prefix}}_{{base32}})"
        )

    @staticmethod
    def parse_guid(guid: str, expected_prefix: str) -> uuid.UUID:
        """
        Parse a GUID string to UUID, validating the prefix.

        Convenience method for service layer lookups.
        Alias for parse_identifier with required prefix.

        Args:
            guid: GUID string
            expected_prefix: Expected entity prefix

        Returns:
            UUID object

        Raises:
            ValueError: If format invalid or prefix doesn't match
        """
        return GuidService.parse_identifier(guid, expected_prefix)


# Backward compatibility alias
ExternalIdService = GuidService
